Improve GitHub status code and update contrib's GitHub webhook

apilloud opened the pull request:
Whole bunch of changes all lumped together here:

1. Updates the structure be closer to the git_buildbot changesource.
2. Adds flag for only reporting the head of a push rather than all commits.
3. Updates fields pulled from GitHub JSON to match the latest APIs
4. Adds support for Pull Requests.
5. Modifies GitHub status reporter to work with GitHub webhook changes by default. 
6. Fixes various bugs.

yetanotherion commented on master/contrib/github_buildbot.py:
```diff
@@ -104,10 +105,10 @@ def render_POST(self, request):
-        # Reject non-push, non-ping events
-        if event_type != "push":
+        # Reject events not involving code changes
+        if event_type not in ("push", "pull_request"):
```
an enum for even_type may be welcome
https://docs.python.org/3/library/enum.html

apilloud replied:
@yetanotherion Enum is new to Python 3.4. Wouldn't that break compatibility with older versions of python (such as 2.7)?

yetanotherion replied:
https://pypi.python.org/pypi/enum ?

2015-08-04 22:57 GMT+02:00 Andrew Pilloud <notifications@github.com>:

> In master/contrib/github_buildbot.py
> <https://github.com/buildbot/buildbot/pull/1801#discussion_r36240773>:
>
> > @@ -104,10 +105,10 @@ def render_POST(self, request):
> >              request.setResponseCode(OK)
> >              return json.dumps({"result": "pong"})
> >
> > -        # Reject non-push, non-ping events
> > -        if event_type != "push":
> > +        # Reject events not involving code changes
> > +        if event_type not in ("push", "pull_request"):
>
> @yetanotherion <https://github.com/yetanotherion> Enum is new to Python
> 3.4. Wouldn't that break compatibility with older versions of python (such
> as 2.7)?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/buildbot/buildbot/pull/1801/files#r36240773>.
>



-- 
Ion

sa2ajj replied:
This kind of event handling does not really allow to easily extend the list of events to handle.  Would you consider implementing something like [this](https://github.com/buildbot/buildbot/blob/master/master/buildbot/www/hooks/github.py#L51)?

apilloud replied:
I like that method, it's a great improvement. Thanks for the suggestion!

yetanotherion commented:
minor comment

apilloud commented:
Rebased to latest passing build and fixed PR build breaks.

sa2ajj commented:
I was a bit confused by the PR title.

Actually, it changes GitHub status code and updates contrib's GitHub webhook.

sa2ajj commented on master/contrib/github_buildbot.py:
```diff
@@ -143,7 +144,27 @@ def render_POST(self, request):
+    def process_change(self, change, branch, repo, repo_url):
+        files = change['added'] + change['removed'] + change['modified']
+        who = ""
+        if 'username' in change['author']:
```
Side note: we seem to have a number of places where similar code is present, it feels to be worth to be extracted to function...

sa2ajj commented on master/contrib/github_buildbot.py:
```diff
@@ -143,7 +144,27 @@ def render_POST(self, request):
+            {'revision': change['id'],
+             'revlink': change['url'],
+             'who': who,
+             'comments': change['message'][:1024],
```
I do not think just trimming message after 1K chars is a good idea, I'd prefer to have an indicator that it was trimmed.  Something like `super useful description ... (trimmed as the length of commit messages exceeds 1024 characters)` (and yes, this will lead to a few characters from the original commit message to be stored).

apilloud replied:
I think this might be a more general bug, as the sqlite database doesn't enforce column length limits. Changed to include a message.

sa2ajj commented on master/contrib/github_buildbot.py:
```diff
@@ -153,28 +174,68 @@ def process_change(self, payload, user, repo, repo_url, project, request):
+
+        m = re.match(r"^refs\/(heads|tags)\/(.+)$", refname)
+        if not m:
+            logging.info("Ignoring refname `%s': Not a branch or tag", refname)
```
`Not a branch or a tag`

sa2ajj commented on master/contrib/github_buildbot.py:
```diff
@@ -153,28 +174,68 @@ def process_change(self, payload, user, repo, repo_url, project, request):
+            logging.info("Ignoring refname `%s': Not a branch or tag", refname)
+            return changes
+
+        branch = m.group(2)
```
this is misleading in the light of the above regular expression: it's either a branch or a tag.

apilloud replied:
How about I call it "refname"?

sa2ajj replied:
sounds good to me: it better reflects the nature of the value.

sa2ajj commented on master/contrib/github_buildbot.py:
```diff
@@ -265,6 +326,10 @@ def setup_options():
                            "perspective broker.",
                       default="change:changepw", dest="auth")
 
+    head_commit_help = ("If set, only trigger builds for commits at head")
```
Is there any reason for introducing `head_commit_help`?

apilloud commented:
Made changes in response to feedback.

apilloud commented:
Found a bug in the fields being used to determine the head commit on a push, fixed.

sa2ajj commented:
We have this nasty problem with certain recent versions of Twisted.  Could you please rebase against master so the tests would show the actual situation?

sa2ajj commented on master/contrib/github_buildbot.py:
```diff
@@ -100,20 +101,16 @@ def render_POST(self, request):
+        if handler is None:
             logging.info(
-                "Rejecting request.  Expected a push event but received %r instead.",
+                "Rejecting request. Received unexpected event %r.",
```
Rather 'unsupported'

sa2ajj commented on master/contrib/github_buildbot.py:
```diff
@@ -153,28 +174,69 @@ def process_change(self, payload, user, repo, repo_url, project, request):
+
+        if payload['action'] not in ("opened", "synchronize"):
+            logging.info("PR %r %r, ignoring",
+                         payload['number'], payload['action'])
```
I think an explicit return would be good here.

sa2ajj commented:
Looks good to me otherwise.

apilloud commented:
Made changes and rebased.

apilloud commented:
Rebased again.

apilloud commented:
Rebased again, cleaned up style issues found by pylint, added tests for github status changes.

sa2ajj commented:
Looks like we have failed tests:
```
[ERROR]
Traceback (most recent call last):
  File "/home/travis/build/buildbot/buildbot/master/buildbot/test/unit/test_status_github.py", line 497, in test_repoOverDefined
    self.assertRaisesConfigError(
exceptions.AttributeError: 'TestGitHubStatus' object has no attribute 'assertRaisesConfigError'

buildbot.test.unit.test_status_github.TestGitHubStatus.test_repoOverDefined
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/travis/build/buildbot/buildbot/master/buildbot/test/unit/test_status_github.py", line 508, in test_repoUnderDefined
    self.assertRaisesConfigError(
exceptions.AttributeError: 'TestGitHubStatus' object has no attribute 'assertRaisesConfigError'

buildbot.test.unit.test_status_github.TestGitHubStatus.test_repoUnderDefined
```

sa2ajj commented:
It looks like you need to add `ConfigErrorsMixin` as, for example, in [here](https://github.com/buildbot/buildbot/blob/917f34b53a5a8706aa4003f065883d93e7b1bec0/master/buildbot/test/unit/test_config.py#L131)

apilloud commented:
The ./common/validate.sh script doesn't appear to run the tests I added. Fixed the test failures, rebased (due to odd failures), and fixed a typo in one of the config error messages.

sa2ajj commented:
Thank you very much!
